bdns/dns.go

```diff
@@ -619,3 +641,48 @@ func logDNSError(
 			underlying)
 	}
 }
+
+type dohExchanger struct {
+	hc http.Client
+}
+
+// Exchange sends a DoH query to the provided DoH server and returns the response.
+func (d *dohExchanger) Exchange(query *dns.Msg, server string) (*dns.Msg, time.Duration, error) {
+	q, err := query.Pack()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// The default Unbound URL template
+	url := fmt.Sprintf("https://%s/dns-query", server)
+	req, err := http.NewRequest("POST", url, strings.NewReader(string(q)))
+	if err != nil {
+		return nil, 0, err
+	}
+	req.Header.Set("Content-Type", "application/dns-message")
+	req.Header.Set("Accept", "application/dns-message")
+
+	start := time.Now()
+	resp, err := d.hc.Do(req)
+	if err != nil {
+		return nil, 0, err
```

Comment:
Should probably return the elapsed time in these error cases too, so that metrics can accurately report how long we took to error out.